src/runtime/crdt/crdt-entity.ts

```diff
@@ -0,0 +1,202 @@
+            op = {type: EntityOpTypes.Remove, removed: operation.value, actor: operation.actor, clock: operation.clock, field: collection};
+          }
+          modelOps.push(op);
+        }
+        for (const operation of collectionChanges[collection].otherChange.operations) {
+          let op;
+          if (operation.type as CollectionOpTypes === CollectionOpTypes.Add) {
+            op = {type: EntityOpTypes.Add, added: operation.value, actor: operation.actor, clock: operation.clock, field: collection};
+          } else {
+            op = {type: EntityOpTypes.Remove, removed: operation.value, actor: operation.actor, clock: operation.clock, field: collection};
+          }
+          otherOps.push(op);
+        }     
+      }
+      return {modelChange: {changeType: ChangeType.Operations, operations: modelOps}, 
+              otherChange: {changeType: ChangeType.Operations, operations: otherOps}};
+    } else {
+      // need to map this.model to get the data out.
+
+      const change: EntityChange<S, C> = {changeType: ChangeType.Model, modelPostChange: this.getData()};
+      return {modelChange: change, otherChange: change};
+    }
+  }
+
+  applyOperation(op: EntityOperation<S, C>): boolean {
+    if (op.type === EntityOpTypes.Set || op.type === EntityOpTypes.Clear) {
+      if (!this.model.singletons[op.field]) {
+        if (this.model.collections[op.field as keyof C]) {
+          throw new Error(`Can't apply ${op.type === EntityOpTypes.Set ? 'Set' : 'Clear'} operation to collection field ${op.field}`);
+        }
+        throw new Error(`Invalid field: ${op.field} does not exist`);
+      }
+    } else {
+      if (!this.model.collections[op.field]) {
+        if (this.model.singletons[op.field as keyof S]) {
+          throw new Error(`Can't apply ${op.type === EntityOpTypes.Add ? 'Add' : 'Remove'} operation to singleton field ${op.field}`);
+        }
+        throw new Error(`Invalid field: ${op.field} does not exist`);
+      }
+    }
+    switch (op.type) {
+      case EntityOpTypes.Set:
+        return this.model.singletons[op.field].applyOperation(
+          {type: SingletonOpTypes.Set, actor: op.actor, value: op.value, clock: op.clock}
+        );
+      case EntityOpTypes.Clear:
+        return this.model.singletons[op.field].applyOperation(
+          {type: SingletonOpTypes.Clear, actor: op.actor, clock: op.clock}
+        );
+      case EntityOpTypes.Add:
+        return this.model.collections[op.field].applyOperation(
+          {type: CollectionOpTypes.Add, actor: op.actor, clock: op.clock, added: op.value}
+        );
+      case EntityOpTypes.Remove:
+        return this.model.collections[op.field].applyOperation(
+          {type: CollectionOpTypes.Remove, actor: op.actor, clock: op.clock, removed: op.value}
+        );
+      default:
+        throw new Error(`Unexpected op ${op} for Entity CRDT`);
+    }
+  }
+
+  getData(): EntityData<S, C> {
+    const singletons = {};
+    const collections = {};
```

Comment:
Types?

Reply:
See above issues; some of them apply here too.